src/enzo/ActiveParticle_SmartStar.C

```diff
@@ -0,0 +1,1058 @@
+    if (i < 0 || i > GridXSize-1 || j < 0 || j > GridYSize-1 || k < 0 || k > GridZSize-1){
+      fprintf(stdout, "Particle out of grid; xind, yind, zind = %"ISYM", %"ISYM", %"ISYM"\n",i,j,k);
+      continue;
+    }
+    // Calculate serial index
+
+    int index = GRIDINDEX_NOGHOST(i,j,k);
+    // If using distributed feedback, check if particle is too close 
+    // to the boundary and adjust indices accordingly
+
+    if (FeedbackDistRadius > 0)
+      {
+	i = max(NumberOfGhostZones + FeedbackDistRadius,
+		min(GridXSize - NumberOfGhostZones - FeedbackDistRadius - 1, i));
+	j = max(NumberOfGhostZones + FeedbackDistRadius,
+		min(GridYSize - NumberOfGhostZones - FeedbackDistRadius - 1, j));
+	k = max(NumberOfGhostZones + FeedbackDistRadius,
+		min(GridZSize - NumberOfGhostZones - FeedbackDistRadius - 1, k));	
+      }
+    // Add energy to the energy field
+    for (int kc = k - FeedbackDistRadius; kc <= k + FeedbackDistRadius; kc++){
+      int stepk = ABS(kc - k);
+      for (int jc = j - FeedbackDistRadius; jc <= j + FeedbackDistRadius; jc++){
+	int stepj = stepk + ABS(jc - j);
+	for (int ic = i - FeedbackDistRadius; ic <= i + FeedbackDistRadius; ic++){
+	  int cellstep = stepj + ABS(ic - i);
+	  int DistIndex = GRIDINDEX_NOGHOST(ic,jc,kc);
+	  if (cellstep <= FeedbackDistCellStep) {
+	    float energybefore = totalenergy[DistIndex];
+	    totalenergy[DistIndex] = ((totalenergy[DistIndex]*density[DistIndex]) + 
+				      LThisTimestep + sn_nrg_thistimestep)/density[DistIndex];
+	    //printf("%s: Accretion Energy added = %e [code energy]\n", __FUNCTION__, LThisTimestep*(StarMass*SolarMass));
+	    //printf("%s: Supernova Energy added = %e [code energy]\n", __FUNCTION__, sn_nrg_thistimestep*(StarMass*SolarMass));
+	    //printf("%s: Fractional Energy increase = %e\n", __FUNCTION__,
+	    // (totalenergy[DistIndex] - energybefore)/energybefore);
+	    if (DualEnergyFormalism == 1) {
+	      float energybefore = gasenergy[DistIndex];
+	      gasenergy[DistIndex] = ((gasenergy[DistIndex]*density[DistIndex]) + 
+				      LThisTimestep)/density[DistIndex];
+	      //printf("%s: Fractional Gas Energy increase = %e\n", __FUNCTION__,
+	      //   (gasenergy[DistIndex] - energybefore)/energybefore);
+	    }
+	  }
+	}
+      }
+    }
+
+  }
+  
+  return SUCCESS;
+}
+
+void ActiveParticleType_SmartStar::DescribeSupplementalData(ActiveParticleFormationDataFlags &flags)
+{
+  flags.DarkMatterDensity = true;
+  flags.Temperature = true;
+  flags.MetalField = true;
+  if(MultiSpecies > 1)
+    flags.H2Fraction = true;
+#if COOLING_TIME
+  flags.CoolingTime = true;
+#else
+  flags.CoolingTime = false;
+#endif
+}
+
+
+template <class active_particle_class>
+int ActiveParticleType_SmartStar::BeforeEvolveLevel
+(HierarchyEntry *Grids[], TopGridData *MetaData,
+ int NumberOfGrids, LevelHierarchyEntry *LevelArray[], 
+ int ThisLevel, bool CallEvolvePhotons, 
+ int TotalStarParticleCountPrevious[],
+ int SmartStarID)
+{
```

Comment:
need some info about this function in context of SmartStar

Reply:
Done